pkg/filesystem/fs/fs.go

```diff
@@ -92,26 +91,38 @@ func NewFileSystem(ctx context.Context, opt ...NewFSOpt) (*Filesystem, error) {
 		//		a new nydusd for it.
 		// TODO: We still need to consider shared daemon the time sequence of initializing daemon,
 		// start daemon commit its state to DB and retrieving its state.
-		if d := fs.getSharedDaemon(); d == nil {
-			log.L.Infof("initializing the shared nydus daemon")
-			if err := fs.initSharedDaemon(); err != nil {
-				return nil, errors.Wrap(err, "start shared nydusd daemon")
-			}
+		log.L.Infof("initializing the shared nydus daemon")
+		if err := fs.initSharedDaemon(); err != nil {
+			return nil, errors.Wrap(err, "start shared nydusd daemon")
 		}
 	}
 
 	// Try to bring all persisted and stopped nydusd up and remount Rafs
-	if len(recoveringDaemons) != 0 {
-		for _, d := range recoveringDaemons {
-			if err := fs.Manager.StartDaemon(d); err != nil {
-				return nil, errors.Wrapf(err, "start daemon %s", d.ID())
-			}
-			if err := d.WaitUntilState(types.DaemonStateRunning); err != nil {
-				return nil, errors.Wrapf(err, "recover daemon %s", d.ID())
-			}
-			if err := d.RecoveredMountInstances(); err != nil {
-				return nil, errors.Wrapf(err, "recover daemons")
-			}
+	for _, d := range recoveringDaemons {
+		if err := fs.Manager.StartDaemon(d); err != nil {
+			return nil, errors.Wrapf(err, "start daemon %s", d.ID())
+		}
+		if err := d.WaitUntilState(types.DaemonStateRunning); err != nil {
+			return nil, errors.Wrapf(err, "wait for daemon %s", d.ID())
+		}
+		if err := d.RecoveredMountInstances(); err != nil {
+			return nil, errors.Wrapf(err, "recover mounts for daemon %s", d.ID())
+		}
+
+		// Found shared daemon
+		// Fscache userspace daemon has no host mountpoint.
+		if d.HostMountpoint() == fs.rootMountpoint || fs.fsDriver == config.FsDriverFscache {
```

Comment:
Maybe put it in a helper func.

Reply:
extracted a help function tryRetainSharedDaemon